macos: initial support for scanning/connecting/write characteristics/notifications

deadprogram opened the pull request:
This PR adds the initial support for macOS with nothing more than scanning implemented.

It is currently WIP so I am making it draft.

deadprogram committed:
macos: starting point for adding macOS support

Signed-off-by: Ron Evans <ron@hybridgroup.com>

deadprogram committed:
macos: able to retrieve some of the info when scanning

Signed-off-by: Ron Evans <ron@hybridgroup.com>

deadprogram committed:
all: use Addresser interface to handle fact that macOS uses UUID instead of MAC as the BLE address for a peripheral

Signed-off-by: Ron Evans <ron@hybridgroup.com>

deadprogram committed:
windows: use Addresser interface for the MAC address for a peripheral

Signed-off-by: Ron Evans <ron@hybridgroup.com>

deadprogram commented:
Latest commit builds and can scan on macOS.

```
make smoketest-macos
# Test on macos.
GOOS=darwin CGO_ENABLED=1 go build -o /tmp/go-build-discard ./examples/scanner

/tmp/go-build-discard
scanning...
found device: ba3cd17f-dbbb-aaba-2f40-e0cffe748605 -94
found device: e8e09032-f1bd-5ebb-874e-11a05eb47579 -95
found device: 1a6b3f14-d497-a284-1e4d-2c6f2d0a0003 -94
found device: 01a84b85-e3eb-589f-ac48-81ed9f578e01 -91
found device: 305648ff-a112-bbab-f648-3e342adad436 -91
found device: 305648ff-a112-bbab-f648-3e342adad436 -91
found device: 44c60243-30f6-d8af-0d47-4a15280d7025 -83
found device: 01a84b85-e3eb-589f-ac48-81ed9f578e01 -91
found device: 01a84b85-e3eb-589f-ac48-81ed9f578e01 -90
found device: 01a84b85-e3eb-589f-ac48-81ed9f578e01 -90 [AV] Samsung Soundbar MS650
found device: 01a84b85-e3eb-589f-ac48-81ed9f578e01 -89
found device: 01a84b85-e3eb-589f-ac48-81ed9f578e01 -89
found device: 01a84b85-e3eb-589f-ac48-81ed9f578e01 -89 [AV] Samsung Soundbar MS650
found device: 21dbaf72-81ca-beb6-a44e-4dc128116156 -93
```

deadprogram committed:
docs: show that we can now scan on macOS

Signed-off-by: Ron Evans <ron@hybridgroup.com>

deadprogram committed:
macos: completed initial implementation

Signed-off-by: Ron Evans <ron@hybridgroup.com>

aykevl commented on README.md:
```diff
@@ -5,17 +5,17 @@
+|                                  | Windows            | Linux              | Nordic chips       | macOS              |
+| -------------------------------- | ------------------ | ------------------ | ------------------ | ------------------ |
+| API used                         | WinRT              | BlueZ (over D-Bus) | SoftDevice         | CoreBluetooth      |
+| Scanning                         | :heavy_check_mark: | :heavy_check_mark: | :heavy_check_mark: | :heavy_check_mark:                |
```
Nit: formatting

deadprogram replied:
Done.

aykevl commented on Makefile:
```diff
@@ -33,3 +33,7 @@ smoketest-linux:
+
+smoketest-macos:
+	# Test on macos.
+	GOOS=darwin CGO_ENABLED=1 go build -o /tmp/go-build-discard ./examples/scanner
```
Would it be possible to add CI tests for this? Probably with a new macOS builder.

deadprogram replied:
Done.

aykevl commented on adapter_darwin.go:
```diff
@@ -0,0 +1,147 @@
+	// peripheral.
+	//prph.SetDelegate(d)
+
+	// Unblock now that we're connected.
+	a.connectChan <- prph
+}
+
+// DidDisconnectPeripheral when peripheral is disconnected.
+func (a *Adapter) DidDisconnectPeripheral(cmgr cbgo.CentralManager, prph cbgo.Peripheral, err error) {
+}
+
+// PeripheralManager delegate functions
+
+// PeripheralManagerDidUpdateState when state updated.
+func (a *Adapter) PeripheralManagerDidUpdateState(pmgr cbgo.PeripheralManager) {
+}
+
+// DidAddService when service added.
+func (a *Adapter) DidAddService(pmgr cbgo.PeripheralManager, svc cbgo.Service, err error) {
+}
+
+// DidStartAdvertising when advertising starts.
+func (a *Adapter) DidStartAdvertising(pmgr cbgo.PeripheralManager, err error) {
+}
+
+// DidReceiveReadRequest when read request received.
+func (a *Adapter) DidReceiveReadRequest(pmgr cbgo.PeripheralManager, cbreq cbgo.ATTRequest) {
+}
+
+// DidReceiveWriteRequests when write requests received.
+func (a *Adapter) DidReceiveWriteRequests(pmgr cbgo.PeripheralManager, cbreqs []cbgo.ATTRequest) {
+}
+
+// CentralDidSubscribe when central subscribed.
+func (a *Adapter) CentralDidSubscribe(pmgr cbgo.PeripheralManager, cent cbgo.Central, cbchr cbgo.Characteristic) {
+}
+
+// CentralDidUnsubscribe when central unsubscribed.
+func (a *Adapter) CentralDidUnsubscribe(pmgr cbgo.PeripheralManager, cent cbgo.Central, chr cbgo.Characteristic) {
+}
```
This code appears to be unused?

deadprogram replied:
This is required to use as a "Peripheral". This is why I have left it in as we will soon need it for the implementation for `Advertisement`.

aykevl commented on adapter_darwin.go:
```diff
@@ -0,0 +1,147 @@
+		serviceUUIDs = append(serviceUUIDs, parsedUUID)
+	}
+
+	// It is never a random address. Is this correct?
```
I have seen random addresses often in the wild, in particular for Nordic chips. But I guess that's irrelevant as it appears that CoreBluetooth doesn't expose MAC addresses.

Note: "random" means the address is generated randomly by the manufacturer and is often fixed for a given chip.

deadprogram replied:
Changed to match how it is working on macOS.

aykevl commented on gap_darwin.go:
```diff
@@ -0,0 +1,135 @@
+	// UUID if this is macOS.
+	UUID
+
+	isRandom bool
```
I assume the `isRandom` bit is included in the UUID and isn't exposed.

deadprogram replied:
Since never "random" per se, changed to show that in the implementation.

aykevl reviewed (commented):
A few initial comments on the draft.

Too bad CoreBluetooth doesn't expose MAC addresses, that makes the whole API less uniform. I had made the (apparently incorrect) assumption that a MAC is so basic that it'll be exposed everywhere.

deadprogram commented:
I have now marked this PR as ready for review. I think it wold be good to get any other changes and then get it merged into `dev` before moving on to additional implementation.

deadprogram committed:
macos: able to discover services and characteristics for a device

Signed-off-by: Ron Evans <ron@hybridgroup.com>

deadprogram commented:
I added the last functionality needed to declare this PR "done" which was to implement `DiscoverServices()` and `DiscoverCharacteristics()` both of which are now complete.

deadprogram commented:
OK, just two more commits now with write characteristics.

aykevl commented:
So, if I understand correctly the methods such as `DidDiscoverPeripheral` are necessary as a callback mechanism?
I'm not so sure about exposing those. They clutter the API, and are really an implementation detail (many functions that a user of the API should never call). Would it be possible to put them on an unexported struct, so that they don't affect the public API?

aykevl commented on gap_darwin.go:
```diff
@@ -0,0 +1,159 @@
+}
+
+// Connect starts a connection attempt to the given peripheral device address.
+func (a *Adapter) Connect(address Addresser, params ConnectionParams) (*Device, error) {
```
It may be a good idea to use some sort of check or mutex, right now it appears that it is not possible to connect to more than one device at once (because of `a.connectChan` for example).

deadprogram replied:
The initial implementation only allows for one device, I was not planning on working on that now.

aykevl commented on gap_darwin.go:
```diff
@@ -0,0 +1,159 @@
+		p.SetDelegate(d)
+		return d, nil
+	case <-time.NewTimer(10 * time.Second).C:
+		return nil, errors.New("timeout on Connect")
```
The documentation says:

> After successfully establishing a local connection to a peripheral, the central manager object calls the centralManager:didConnectPeripheral: method of its delegate object. If the connection attempt fails, the central manager object calls the centralManager:didFailToConnectPeripheral:error: method of its delegate object instead. Attempts to connect to a peripheral don’t time out.

Ideally we'd use these instead of laying our own error handling on top of this.
(Of course, `didFailToConnectPeripheral` is secretly just a timeout but it's a timeout on a much lower level).

deadprogram replied:
This was just a simple implementation, but can probably connect that easily enough.

aykevl commented on gap_darwin.go:
```diff
@@ -0,0 +1,159 @@
+
+// Peripheral returns the Device's cbgo.Peripheral
+func (d *Device) Peripheral() (prph cbgo.Peripheral) {
+	return d.prph
```
Is there a reason to expose this?
I'm personally a bit hesitant with exposing things that look like implementation details. If API users need access to the `prph` member I'd rather just make it exported (`Prph`) instead.

deadprogram replied:
It is needed to be called from the child of a child, hence exposing it. I can change to exported member.

aykevl replied:
I'm not sure I understand. If it is only needed in the bluetooth package, you can simply use it as is, without exporting?

deadprogram replied:
Good point. Changed.

aykevl commented on gap_darwin.go:
```diff
@@ -0,0 +1,159 @@
+// CharsChan returns the Device's charsChan channel used for
+// blocking on discovering the characteristics for a service.
+func (d *Device) CharsChan() chan error {
+	return d.charsChan
```
Even more with this channel: reading from it can mess with package internal state so I think it should remain unexported.

deadprogram replied:
It was the least complicated way I could figure out how to use the CB delegation in `Device` from `DeviceService`.

deadprogram replied:
Changed to private implementation.

aykevl commented on gap_linux.go:
```diff
@@ -11,6 +11,29 @@ import (
 
+// Address contains a Bluetooth address, which is a MAC address plus some extra
+// information.
+type Address struct {
```
I would propose to rename this to something like `MACAddress` and share it between all implementations (in gap.go). That avoids some duplication.

deadprogram replied:
OK, can do that.

deadprogram replied:
Done.

deadprogram commented:
In order to take advantage of a bunch of the predemined functionality by using https://github.com/tinygo-org/bluetooth/blob/macos/adapter_darwin.go#L12 you need to implement those methods. I think without this there would be a lot more of them! macOS is macOS, we just do what they say.

aykevl commented on gattc_darwin.go:
```diff
@@ -0,0 +1,121 @@
+			svcs = append(svcs, svc)
+		}
+		return svcs, nil
+	case <-time.NewTimer(10 * time.Second).C:
+		return nil, errors.New("timeout on DiscoverServices")
```
I suspect this doesn't really time out either, it either succeeds or the entire connection is broken.

deadprogram replied:
Same as above comment.

aykevl commented on gattc_darwin.go:
```diff
@@ -0,0 +1,121 @@
+			chars = append(chars, char)
+		}
+		return chars, nil
+	case <-time.NewTimer(10 * time.Second).C:
+		return nil, errors.New("timeout on DiscoverCharacteristics")
```
Same here.

deadprogram replied:
Same as my previous comment.

aykevl commented on gattc_darwin.go:
```diff
@@ -0,0 +1,121 @@
+		return errors.New("must provide a callback for EnableNotifications")
+	}
+
+	return nil
```
Is there a reason to leave this function unimplemented?

deadprogram replied:
Because I am working on it, and left it unimplemented for now.

deadprogram replied:
This has now been implemented.

aykevl commented:
I see. However, they don't need to be on the `Adapter` type I think. They would be hidden from the user with something like this:

```go
type Adapter struct {
    delegate centralManagerDelegate
}

type centralManagerDelegate struct {
    cbgo.CentralManagerDelegateBase
    a *Adapter
}


// CentralManagerDidUpdateState when central manager state updated.
func (delegate *centralManagerDelegate) CentralManagerDidUpdateState(cmgr cbgo.CentralManager) {
	// powered on?
	if cmgr.State() == cbgo.ManagerStatePoweredOn {
		close(delegate.a.poweredChan)
	}

	// TODO: handle other state changes.
}
```

This provides the same functionality without cluttering the API.

deadprogram commented:
I discovered I can remove a bunch of the implementation that is not being used, coming in future commit. Also, I was trying to have a simpler implementation with less indirection, since there is already a lot of complexity in the macOS CB code itself.

deadprogram committed:
gap: switch to use MACAddress struct when possible for shared implementation

Signed-off-by: Ron Evans <ron@hybridgroup.com>

deadprogram committed:
macos: remove unneeded functions to export internal implementaions.

Signed-off-by: Ron Evans <ron@hybridgroup.com>

deadprogram commented:
Added some delegate types, and have otherwise handled all outstanding feedback items, I think.

deadprogram committed:
macos: added characteristic notifications

Signed-off-by: Ron Evans <ron@hybridgroup.com>

deadprogram commented:
Added characteristic notifications.  No more features in this PR, I mean it! :smiley_cat:

deadprogram committed:
gap: correct use of Address on Linux platform

Signed-off-by: deadprogram <ron@hybridgroup.com>

deadprogram commented:
I squashed the commits in this branch down based on the functionality, and rebased against `dev`. Also, have tested on macOS and Linux plus Circuit Playground Bluefruit.

Basically, we can now finally ready to merge! Thanks for all your help on this @aykevl

deadprogram commented:
@aykevl any more feedback on this PR?

aykevl commented:
I'll try to get to this PR tomorrow. I was a bit busy with unrelated things (which is now mostly done).

aykevl commented on adapter_darwin.go:
```diff
@@ -0,0 +1,123 @@
+		serviceUUIDs = append(serviceUUIDs, parsedUUID)
+	}
+
+	// It is never a random address on macOS.
```
Nit: this comment seems to be outdated (there is no way of knowing whether an address is random as not even the MAC address itself is exposed).

deadprogram replied:
Updated.

aykevl reviewed (approved):
Looks good to me. Feel free to merge, unless you want to fix the nit.

deadprogram committed:
docs: better explanation of peripheral UUID on macOS

Signed-off-by: deadprogram <ron@hybridgroup.com>

deadprogram commented:
Now merging, thank you very much @aykevl for your help landing this PR!
